### ISSUE TITLE
www.extensionworkshop.com

### ISSUE THREAD

**Issue by willdurand** [mozilla/extension-workshop#988]
### Describe the problem and steps to reproduce it:

1. Go to https://www.extensionworkshop.com/

### What happened?

The EW is served.

### What did you expect to happen?

I am not sure but shouldn't we send a redirect instead? (to the domain without `www`). This might not be an issue after all but I didn't find anything in the GH issues.

**Comment by muffinresearch** [comment]
Thanks for filing this, it's not been noticed before. 

@bqbn since this is in S3, is there anything we can do with this via cloudfront configuration?

**Comment by bqbn** [comment]
I'm trying to understand, are you saying `www.extensionworkshop.com` should be redirected to `extensionworkshop.com`? If so, why?

**Comment by muffinresearch** [comment]
> I'm trying to understand, are you saying `www.extensionworkshop.com` should be redirected to `extensionworkshop.com`? If so, why?

The canonical domain has no www, so it's conventional to redirect the www to the non-www.

**Comment by bqbn** [comment]
I haven't looked into the tech details to see if the redirection is doable or not. However, other major sites seem to do the opposite of what we are trying to do, i.e. they redirect the non-www name to the www name. Examples include `google.com -> www.google.com`, `mozilla.org -> www.mozilla.org`, `facebook.com -> www.facebook.com` and etc.

Do we want to do the same, if the redirection can be implemented via Cloudfront configuration?

**Comment by muffinresearch** [comment]
Since we've always linked the non-www we'd likely be best off sticking with the non-www version. This [link documents some of the main points related to the considerations around www or non-www domains](https://www.spyfu.com/blog/www-vs-non-www/). 

Give this is a static docs site, I don's see that the cons illustrated on that link apply in this case. But if you think there are other reasons why we should reconsider please fee free to voice them.

**Comment by bqbn** [comment]
I personally think `non-www -> www` makes more sense but I don't have a strong opinion in this case.

And after doing some research, I think redirection is doable. However, it can't be done without adding additional AWS resources. We'll probably implement it by using a second CloudFront distribution and a second S3 bucket to host the www domain and do the redirection.

Could you comment whether it is still worth doing it given the added costs and complexity of the system design? The cost may not be too much though since redirection doesn't transmit too many bytes.

**Comment by muffinresearch** [comment]
> I personally think `non-www -> www` makes more sense but I don't have a strong opinion in this case.
> 
> And after doing some research, I think redirection is doable. However, it can't be done without adding additional AWS resources. We'll probably implement it by using a second CloudFront distribution and a second S3 bucket to host the www domain and do the redirection.
> 
> Could you comment whether it is still worth doing it given the added costs and complexity of the system design? The cost may not be too much though since redirection doesn't transmit too many bytes.

Would it make a difference putting nginx in front of S3? Since I think we are going to need that if we want to fix https://github.com/mozilla/extension-workshop/issues/935

**Comment by bqbn** [comment]
> Would it make a difference putting nginx in front of S3?

Cost wise, it's more expensive, because we will need at least 2 EC2 instances for redundancy and 1 ELB for load balancing.

 > Since I think we are going to need that if we want to fix #935

As I noted in https://github.com/mozilla/extension-workshop/issues/935#issuecomment-799796523, a second option is to leverage the Lambda@Edge to append a slash to the path, and is preferable.

**Comment by bqbn** [comment]
Also, how important is it to fix this issue? Is it OK for the site to not have the redirection?

If the stake is high and fixing it is important, then I suggest that we put in a second cloudfront+S3 pair and have it addressed.

If without the redirection doesn't have much

**Comment by willdurand** [comment]
> Also, how important is it to fix this issue? Is it OK for the site to not have the redirection?

I think for SEO, it's a bit better if we have the redirect.

**Comment by bqbn** [comment]
I set up the redirection in -dev and -stage for now. Please test them and if it works as expected, we can set it up in -prod.

* In -dev, it is `www-extensionworkshop.dev.mozaws.net` redirecting to `extensionworkshop-dev.allizom.org`.
* In -stage, it is `www-extensionworkshop.stage.mozaws.net` redirecting to `extensionworkshop.allizom.org`.

**Comment by willdurand** [comment]
> I set up the redirection in -dev and -stage for now. Please test them and if it works as expected, we can set it up in -prod.
> 
>     * In -dev, it is `www-extensionworkshop.dev.mozaws.net` redirecting to `extensionworkshop-dev.allizom.org`.
> 
>     * In -stage, it is `www-extensionworkshop.stage.mozaws.net` redirecting to `extensionworkshop.allizom.org`.


These redirections aren't really similar to the one we need for -prod so I am not sure if they are that useful.

It works with `curl`, though:

```
$ curl -I https://www-extensionworkshop.dev.mozaws.net/
HTTP/1.1 301 Moved Permanently
Content-Length: 0
Connection: keep-alive
Date: Tue, 25 May 2021 16:32:24 GMT
Location: https://extensionworkshop-dev.allizom.org/
Server: AmazonS3
X-Cache: Miss from cloudfront
Via: 1.1 e0efba8a72628bfc3dc6d4d637b28302.cloudfront.net (CloudFront)
X-Amz-Cf-Pop: FRA2-C1
X-Amz-Cf-Id: gnumKji_2sO1_CXotqH55XNHgEIrRiR8eo3g3HlmSfc4aq6AaRyRCw==
```

**Comment by bqbn** [comment]
> These redirections aren't really similar to the one we need for -prod so I am not sure if they are that useful.

That's right. This is to demonstrate the idea that we can use a separate pair of Cloudfront distribution and  S3 to implement the redirection. In -prod, we'll redirect the `www` domain to the non-www domain as requested.

Would you agree to push the change to -prod?

**Comment by willdurand** [comment]
> Would you agree to push the change to -prod?

Yes.

**Comment by muffinresearch** [comment]
@bqbn should the redirect be deployed? I don't see it, I just see the served site on both www and non-www.

**Comment by bqbn** [comment]
Sorry this slipped the cracks. The redirection is now deployed.

```
curl -I https://www.extensionworkshop.com/
HTTP/1.1 301 Moved Permanently
Content-Length: 0
Connection: keep-alive
Date: Mon, 22 Nov 2021 19:41:12 GMT
Location: https://extensionworkshop.com/
Server: AmazonS3
X-Cache: Miss from cloudfront
Via: 1.1 87136170926d082ce5ff23d5ad5be32c.cloudfront.net (CloudFront)
X-Amz-Cf-Pop: YVR50-C1
X-Amz-Cf-Id: EHGW_SOOGc-CvoVTQtjvcrJNfdsRghiAuTNUkAGujFLnKuiW4Q7ifA==
```
